tsl/src/bgw_policy/job_api.c

```diff
@@ -0,0 +1,233 @@
+
+	if (pg_proc_aclcheck(proc, owner, ACL_EXECUTE) != ACLCHECK_OK)
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 errmsg("permission denied for function \"%s\"", func_name)));
+
+	/* Verify that the owner can create a background worker */
+	ts_bgw_job_validate_job_owner(owner, JOB_TYPE_CUSTOM);
+
+	/* Next, insert a new job into jobs table */
+	namestrcpy(&application_name, "Custom Job");
+	namestrcpy(&custom_name, "custom");
+	namestrcpy(&proc_schema, get_namespace_name(get_func_namespace(proc)));
+	namestrcpy(&proc_name, func_name);
+	namestrcpy(&owner_name, GetUserNameFromId(owner, false));
+
+	job_id = ts_bgw_job_insert_relation(&application_name,
+										&custom_name,
+										schedule_interval,
+										&max_runtime,
+										DEFAULT_MAX_RETRIES,
+										&retry_period,
+										&proc_name,
+										&proc_schema,
+										&owner_name,
+										scheduled,
+										0,
+										config);
+
+	if (!PG_ARGISNULL(3))
+	{
+		TimestampTz initial_start = PG_GETARG_TIMESTAMPTZ(3);
+		ts_bgw_job_stat_upsert_next_start(job_id, initial_start);
+	}
+
+	PG_RETURN_INT32(job_id);
+}
+
+/*
+ * CREATE OR REPLACE FUNCTION delete_job(job_id INTEGER) RETURNS VOID
+ */
+Datum
+job_delete(PG_FUNCTION_ARGS)
+{
+	int32 job_id = PG_GETARG_INT32(0);
+	BgwJob *job = ts_bgw_job_find(job_id, CurrentMemoryContext, true);
+	Oid owner = get_role_oid(NameStr(job->fd.owner), false);
+
+	if (!has_privs_of_role(GetUserId(), owner))
+		ereport(ERROR,
+				(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				 errmsg("insufficient permissions to delete job for user \"%s\"",
+						NameStr(job->fd.owner))));
+
+	ts_bgw_job_delete_by_id(job_id);
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * CREATE OR REPLACE PROCEDURE run_job(job_id INTEGER)
+ */
+Datum
+job_run(PG_FUNCTION_ARGS)
+{
+	int32 job_id = PG_GETARG_INT32(0);
+	BgwJob *job = ts_bgw_job_find(job_id, CurrentMemoryContext, true);
+
+	tsl_bgw_policy_job_execute(job);
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * CREATE OR REPLACE FUNCTION alter_job(
+ * 0    job_id INTEGER,
+ * 1    schedule_interval INTERVAL = NULL,
+ * 2    max_runtime INTERVAL = NULL,
+ * 3    max_retries INTEGER = NULL,
+ * 4    retry_period INTERVAL = NULL,
+ * 5    scheduled BOOL = NULL,
+ * 6    config JSONB = NULL,
+ * 7    next_start TIMESTAMPTZ = NULL
+ * 8    if_exists BOOL = FALSE,
+ * ) RETURNS TABLE (
+ *      job_id INTEGER,
+ *      schedule_interval INTERVAL,
+ *      max_runtime INTERVAL,
+ *      max_retries INTEGER,
+ *      retry_period INTERVAL,
+ *      scheduled BOOL,
+ *      config JSONB,
+ *      next_start TIMESTAMPTZ
+ * )
+ */
+Datum
+job_alter(PG_FUNCTION_ARGS)
+{
+	BgwJob *job;
+	BgwJobStat *stat;
+	TupleDesc tupdesc;
+	Datum values[ALTER_JOB_NUM_COLS];
+	bool nulls[ALTER_JOB_NUM_COLS] = { false };
+	HeapTuple tuple;
+	TimestampTz next_start;
+
+	int job_id = PG_GETARG_INT32(0);
+	bool if_exists = PG_GETARG_BOOL(8);
+
+	/* First get the job */
+	job = ts_bgw_job_find(job_id, CurrentMemoryContext, false);
```

Comment:
Comment just duplicates code.
```suggestion
	job = ts_bgw_job_find(job_id, CurrentMemoryContext, false);
```